spec.html

```diff
@@ -32497,173 +32497,107 @@ <h1>String.prototype.replace ( _searchValue_, _replaceValue_ )</h1>
           1. Else,
             1. Assert: Type(_replaceValue_) is String.
             1. Let _captures_ be a new empty List.
             1. Let _replacement_ be ! GetSubstitution(_searchString_, _string_, _position_, _captures_, *undefined*, _replaceValue_).
-          1. Return the string-concatenation of _preserved_, _replacement_, and the substring of _string_ from _position_ + _searchLength_.
+          1. Return the string-concatenation of _preceding_, _replacement_, and _following_.
         </emu-alg>
         <emu-note>
           <p>The `replace` function is intentionally generic; it does not require that its *this* value be a String object. Therefore, it can be transferred to other kinds of objects for use as a method.</p>
         </emu-note>
 
-        <emu-clause id="sec-getsubstitution" type="abstract operation">
+        <emu-clause id="sec-getsubstitution" type="abstract operation" oldids="table-replacement-text-symbol-substitutions,table-45">
           <h1>
             GetSubstitution (
               _matched_: unknown,
               _str_: unknown,
               _position_: a non-negative integer,
               _captures_: unknown,
               _namedCaptures_: unknown,
-              _replacement_: unknown,
+              _replacer_: unknown,
             )
           </h1>
           <dl class="header">
           </dl>
+          <p>For the purposes of this abstract operation, a <em>decimal digit</em> is a code unit in the range 0x0030 (DIGIT ZERO) to 0x0039 (DIGIT NINE) inclusive.</p>
           <emu-alg>
             1. Assert: Type(_matched_) is String.
             1. Let _matchLength_ be the number of code units in _matched_.
             1. Assert: Type(_str_) is String.
             1. Let _stringLength_ be the number of code units in _str_.
             1. Assert: _position_ &le; _stringLength_.
-            1. Assert: _captures_ is a possibly empty List of Strings.
-            1. Assert: Type(_replacement_) is String.
+            1. Assert: _captures_ is a possibly empty List, each of whose elements is a String or *undefined*.
+            1. Assert: Type(_replacer_) is String.
+            1. Let _preceding_ be the substring of _str_ from 0 to _position_.
             1. Let _tailPos_ be _position_ + _matchLength_.
-            1. Let _m_ be the number of elements in _captures_.
-            1. Let _result_ be the String value derived from _replacement_ by copying code unit elements from _replacement_ to _result_ while performing replacements as specified in <emu-xref href="#table-replacement-text-symbol-substitutions"></emu-xref>. These `$` replacements are done left-to-right, and, once such a replacement is performed, the new replacement text is not subject to further replacements.
-            1. Return _result_.
+            1. If _tailPos_ &ge; _stringLength_, then
```

Comment:
Not related to your PR, but this is a dumb case. Assuming I understand this correctly, it only happens when code has overridden the default behavior of `exec` in such a way that it claims to have find a matching substring at a position where that substring is longer than the number of characters left following that position in the input string.

That might warrant a note (though not necessarily in this PR).

Reply:
Yeah, it struck me as odd too. I *think* that's the only way that `_tailPos_ > _stringLength_` can happen.

Reply:
I believe have that branch covered in es-abstract's GetSubstitution with this invocation:
```js
GetSubstitution('abcdef', 'abcdefghi', 0, [], '>$`<'),
```

does that help?

Reply:
@ljharb With that invocation, you should have `matchLength = 'abcdef'.length = 6` and `stringLength =  'abcdefghi'.length = 9`, and hence `tailPos = position + matchLength = 0 + 6 = 6`, and hence `tailPos < stringLength`. So I don't think that invocation should cover the `_tailPos_ > _stringLength_` case I'm referring to here.

Reply:
ah, i may have grabbed the wrong test case.
```js
GetSubstitution('def', 'abcdefghi', 3, [], '>$`<'),
```
?

If that's still wrong, I'd love help triggering it :-)

Reply:
Same deal with that one:

`position = 3`, `matchLength = 'def'.length = 3`, `stringLength = 'abcdefghi'.length = 9`, and hence `tailPos = position + matchLength = 3 + 3 = 6`, and hence `tailPos < stringLength`. (Oh, and in both cases the replacement-template-string needs to contain `$'`, not `` $` ``, to be in the relevant row/branch at all.)

The cases which actually exercise it look like `GetSubstitution("1234567", "abc", 0, [], "$'")` or `GetSubstitution("x", "abc", 3, [], "$'")` - in the former case by having the match be longer than the input string, in the latter by having the match be nonempty and occur at the end of the input string.

Obviously that's a very strange thing to do, but you can trigger it by overriding `.exec`, as in

```js
let evil = new RegExp;
evil.exec = () => ({ 0: '1234567', length: 1, index: 0 });
'abc'.replace(evil, `$'`);
```
or

```js
let evil = new RegExp;
evil.exec = () => ({ 0: 'x', length: 1, index: 3 });
'abc'.replace(evil, `$'`);
```

(Fun fact: I just tested these snippets and found bugs in four different JS engines: [SpiderMonkey](https://bugzilla.mozilla.org/show_bug.cgi?id=1724618), [Chakra](https://github.com/chakra-core/ChakraCore/issues/6736), [XS](https://github.com/Moddable-OpenSource/moddable/issues/680), and [GraalJS](https://github.com/oracle/graaljs/issues/488).)

Reply:
(I haven't added a Note yet.)